xcdat/averager.py

```diff
@@ -0,0 +1,543 @@
+    lat_bounds : Optional[RegionAxisBounds], optional
+        Tuple of latitude boundaries for regional selection, by default None.
+    lon_bounds : Optional[RegionAxisBounds], optional
+        Tuple of longitude boundaries for regional selection, by default None.
+    is_weighted : bool, True
+        Weighting option to calculate weights proportional to area
+        (True), uniform weights (False).
+
+    Returns
+    -------
+    AxisWeights
+        Set of weights for each eligible dimension.
+    """
+    if is_weighted:
+        weights = _get_area_weights(xvariable, lat_bounds, lon_bounds)
+    else:
+        weights = _get_uniform_weights(xvariable)
+    return weights
+
+
+def _get_area_weights(
+    xvariable: xr.DataArray,
+    lat_bounds: Optional[RegionAxisBounds],
+    lon_bounds: Optional[RegionAxisBounds],
+):
+    """Gets area weights for a dimension with weighted bounds.
+
+    # TODO: Add more description on how weights are generated.
+
+    Parameters
+    ----------
+    xvariable : xr.DataArray
+        xr.DataArray produced using ``open_variable``.
+    lat_bounds : Optional[RegionAxisBounds], optional
+        Tuple of latitude boundaries for regional selection, by default None.
+    lon_bounds : Optional[RegionAxisBounds], optional
+        Tuple of longitude boundaries for regional selection, by default None.
+
+    Returns
+    -------
+    AxisWeights
+        Set of weights for each eligible dimension.
+
+    Notes
+    -----
+    This is a prototype and needs work to make it work more generically. It
+    currently only works with lat / lon. It might be useful to check the
+    latitude axis units.
+    """
+    BoundsByType = TypedDict(
+        "BoundsByType",
+        {"domain": Optional[xr.DataArray], "region": Optional[RegionAxisBounds]},
+    )
+    bounds: Dict[str, BoundsByType] = {
+        "lat": {"domain": xvariable.bounds.get_bounds("lat"), "region": lat_bounds},
+        "lon": {"domain": xvariable.bounds.get_bounds("lon"), "region": lon_bounds},
+    }
+    weights: AxisWeights = {}
+
+    for dim, bounds_by_type in bounds.items():
+        domain_bounds = bounds_by_type["domain"]
+        region_bounds = bounds_by_type["region"]
+
+        if domain_bounds is not None:
+            domain_bounds_copy = domain_bounds.copy()
+
+            if region_bounds is not None:
+                domain_bounds_copy = _scale_dim_to_region(
+                    domain_bounds_copy, region_bounds
+                )
+
+            if dim == "lat":
+                domain_bounds_copy = np.sin(np.radians(domain_bounds_copy))
+
+            # weights are generally the difference of the boundaries
+            # REVIEW: Is this always true for our purposes?
+            dim_wts = np.abs(domain_bounds_copy[:, 1] - domain_bounds_copy[:, 0])
+            dim_wts.attrs = domain_bounds_copy.attrs
+            dim_wts.name = dim + "_wts"
+            weights[dim] = dim_wts
+
+    return weights
```

Comment:
I've updated the logic of `_get_area_weights()`.

It now stores the domain and region bounds in the `bounds` dict, instead of using `xvariable.bounds.bounds` (which gives a bounds dict for CF convention coordinate keys and axis).

I've statically defined the latitude key to "lat" and use `xvariable.bounds.get_bounds("lat"). You'll notice I check for `if dim == "lat"` now instead of `if dim in ["lat", "latitude"]`, since we loop over the `bounds` dict.

Reply:
I realized `xvariable.bounds.get_bounds("lat")` will throw and error if the bounds don't exist.
Instead, I will do:
```
domain_bounds = xvariable.bounds.bounds
bounds: Dict[str, BoundsByType] = {
    "lat": {"domain": bounds.get("lat"), "region": lat_bounds},
    "lon": {"domain": bounds.get("lon"), "region": lon_bounds}, 
}
```